syntax = "proto3";

package key_value_store;

service KeyValueStore {
    // Establece el valor de la clave dada
    rpc Set(SetKeyValue) returns (SetKeyValueResponse);

    // Devuelve el valor de la clave dada
    rpc Get(GetValue) returns (GetValueResponse);

    // Devuelve una lista de valores cuyas claves empiezan por prefixKey.
    rpc GetPrefixKey(GetPrefix) returns (GetPrefixResponse);

    rpc Stat(StatRequest) returns (StatResponse);
}

message SetKeyValue {
    string key = 1; 
    string value = 2;
}

message SetKeyValueResponse {
    bool status = 1;
    string message = 2;
}

message GetValue {
    string key = 1;
}

message GetValueResponse {
    bool status = 1;
    string value = 2; 
}

message GetPrefix {
    string prefixKey = 1;
}

message GetPrefixResponse {
    repeated string keys = 1;
    repeated string values = 2;
}

message StatRequest { }

message StatResponse {
    string time_started = 1;
    int64 total_requests = 2;
    int64 total_set_requests = 3;
    int64 total_get_requests = 4;
    int64 total_get_prefix_requests = 5;
}